clippy_lints/src/unnecessary_indexing.rs

```diff
@@ -0,0 +1,220 @@
+                        )];
+                        index_accesses.extend(r.extra_exprs_copy.iter().map(|x| (x.span, "element".to_owned())));
+
+                        diag.multipart_suggestion(
+                            "consider using if..let syntax",
+                            index_accesses,
+                            Applicability::Unspecified,
+                        );
+                    } else {
+                        let mut index_accesses = vec![(
+                            if_expr.cond.span,
+                            format!("let Some(element) = {}.first()", snippet(cx, receiver.span, "..")),
+                        )];
+                        index_accesses.extend(r.extra_exprs_borrow.iter().map(|x| (x.span, "element".to_owned())));
+                        index_accesses.extend(r.extra_exprs_copy.iter().map(|x| (x.span, "element".to_owned())));
+
+                        diag.multipart_suggestion(
+                            "consider using if..let syntax (variable may need to be dereferenced)",
+                            index_accesses,
+                            Applicability::Unspecified,
+                        );
+                    }
+                },
+            );
+        }
+    }
+}
+
+struct IndexCheckResult<'a> {
+    // the receiver for the index operation, only Some in the event the indexing is via a direct primitive
+    index_receiver: Option<&'a Expr<'a>>,
+    // first local in the block - used as pattern for `Some(pat)`
+    first_local: Option<&'a LetStmt<'a>>,
+    // any other index expressions to replace with `pat` (or "element" if no local exists)
+    extra_exprs_borrow: Vec<&'a Expr<'a>>,
+    // copied extra index expressions: if we only have these and no borrows we can provide a correct suggestion of `let
+    // Some(&a) = ...`
+    extra_exprs_copy: Vec<&'a Expr<'a>>,
+}
+
+/// Checks the block for any indexing of the conditional receiver. Returns `None` if the block
+/// contains code that invalidates the lint, e.g., the receiver is accessed via a mutable reference.
+fn process_indexing<'a>(
+    cx: &LateContext<'a>,
+    block: &'a Block<'a>,
+    conditional_receiver_hid: HirId,
+) -> Option<IndexCheckResult<'a>> {
+    let mut index_receiver: Option<&Expr<'_>> = None;
+    let mut first_local: Option<&LetStmt<'_>> = None;
+    let mut extra_exprs_borrow: Vec<&Expr<'_>> = vec![];
+    let mut extra_exprs_copy: Vec<&Expr<'_>> = vec![];
+
+    // if res == Some(()), then mutation occurred
+    // & therefore we should not lint on this
+    let res = for_each_expr(cx, block.stmts, |x| {
+        let adjustments = cx.typeck_results().expr_adjustments(x);
+        if let ExprKind::Index(receiver, index, _) = x.kind
+            && let ExprKind::Lit(lit) = index.kind
+            && let LitKind::Int(val, _) = lit.node
+            && path_to_local_id(receiver, conditional_receiver_hid)
+            && val.0 == 0
+        {
+            index_receiver = Some(receiver);
```

Comment:
As a callback to https://github.com/rust-lang/rust-clippy/pull/12464#discussion_r1626571462 we could make this a bool and check it before returning to ditch the field